docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+bitmap, if it has one or more null values.
+
+Physically, a struct type has one child array for each field. The
+child arrays are independent and need not be adjacent to each other in
+memory.
+
+For example, the struct (field names shown here as strings for illustration
+purposes)::
+
+    Struct <
+      name: String (= List<char>),
+      age: Int32
+    >
+
+has two child arrays, one ``List<char>`` array (layout as above) and one 4-byte
+primitive value array having ``Int32`` logical type.
+
+**Example Layout: ``Struct<List<char>, Int32>``**
+
+The layout for ``[{'joe', 1}, {null, 2}, null, {'mark', 4}]`` would be: ::
+
+    * Length: 4, Null count: 1
+    * Null bitmap buffer:
+
+      |Byte 0 (validity bitmap) | Bytes 1-63            |
+      |-------------------------|-----------------------|
+      | 00001011                | 0 (padding)           |
+
+    * Children arrays:
+      * field-0 array (`List<char>`):
+        * Length: 4, Null count: 2
+        * Null bitmap buffer:
+
+          | Byte 0 (validity bitmap) | Bytes 1-63            |
+          |--------------------------|-----------------------|
+          | 00001001                 | 0 (padding)           |
+
+        * Offsets buffer:
+
+          | Bytes 0-19     |
+          |----------------|
+          | 0, 3, 3, 3, 7  |
+
+         * Values array:
+            * Length: 7, Null count: 0
+            * Null bitmap buffer: Not required
+
+            * Value buffer:
+
+              | Bytes 0-6      |
+              |----------------|
+              | joemark        |
+
+      * field-1 array (int32 array):
+        * Length: 4, Null count: 1
+        * Null bitmap buffer:
+
+          | Byte 0 (validity bitmap) | Bytes 1-63            |
+          |--------------------------|-----------------------|
+          | 00001011                 | 0 (padding)           |
+
+        * Value Buffer:
+
+          |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-63 |
+          |------------|-------------|-------------|-------------|-------------|
+          | 1          | 2           | unspecified | 4           | unspecified |
+
+While a struct does not have physical storage for each of its semantic
+slots (i.e. each scalar C-like struct), an entire struct slot can be
+set to null via the null bitmap. Any of the child field arrays can
+have null values according to their respective independent null
+bitmaps. This implies that for a particular struct slot the null
+bitmap for the struct array might indicate a null slot when one or
+more of its child arrays has a non-null value in their corresponding
+slot.  When reading the struct array the parent null bitmap is
+authoritative.  This is illustrated in the example above, the child
+arrays have valid entries for the null struct but are 'hidden' from
+the consumer by the parent array's null bitmap.  However, when treated
+independently corresponding values of the children array will be
+non-null.
+
+Dense Union
```

Comment:
Should there be a "Union" paragraph heading and then move the "Dense Union" sub-heading just before the dense union-specific part?